certificate_loader.go

```diff
@@ -17,113 +17,192 @@
+func (c CertificateLoader) certFiles() ([]string, error) {
+	var files []string
+
+	if _, err := os.Stat(c.CertFile); err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("unable to stat %s\n%w", c.CertFile, err)
+	} else if err == nil {
+		files = append(files, c.CertFile)
+	}
+
+	re := regexp.MustCompile(`^[[:xdigit:]]{8}\.[\d]$`)
+	for _, d := range c.CertDirs {
+		c, err := ioutil.ReadDir(d)
+		if os.IsNotExist(err) {
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("unable to list children of %s\n%w", d, err)
+		}
+
+		for _, c := range c {
+			if c.IsDir() || !re.MatchString(c.Name()) {
+				continue
+			}
+
+			files = append(files, filepath.Join(d, c.Name()))
+		}
 	}
 
+	return files, nil
+}
+
+func (c CertificateLoader) readBlocks(path string) ([]*pem.Block, error) {
 	var (
 		block  *pem.Block
 		blocks []*pem.Block
 	)
-	for len(rest) != 0 {
+
+	rest, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read %s\n%w", path, err)
+	}
+
+	for {
 		block, rest = pem.Decode(rest)
+		if block == nil {
+			break
+		}
 		blocks = append(blocks, block)
 	}
 
 	return blocks, nil
 }
 
-func (c CertificateLoader) ReadKeyStore() (keystore.KeyStore, error) {
-	in, err := os.Open(c.KeyStorePath)
+func (CertificateLoader) readKeyStore(path string, password string) (keystore.KeyStore, error) {
+	in, err := os.Open(path)
 	if err != nil {
-		return nil, fmt.Errorf("unable to open %s\n%w", c.KeyStorePath, err)
+		return nil, fmt.Errorf("unable to open %s\n%w", path, err)
 	}
 	defer in.Close()
 
-	ks, err := keystore.Decode(in, []byte(c.KeyStorePassword))
+	ks, err := keystore.Decode(in, []byte(password))
 	if err != nil {
 		return nil, fmt.Errorf("unable to decode keystore\n %w", err)
 	}
 
 	return ks, nil
 }
 
-func (c CertificateLoader) WriteKeyStore(ks keystore.KeyStore) error {
-	if unix.Access(c.KeyStorePath, unix.W_OK) != nil {
-		_, _ = fmt.Fprintf(c.Logger, "WARNING: Unable to add container CA certificates to JVM because %s is read-only", c.KeyStorePath)
+func (c CertificateLoader) writeKeyStore(ks keystore.KeyStore, path string, password string) error {
+	if unix.Access(path, unix.W_OK) != nil {
+		_, _ = fmt.Fprintf(c.Logger, "WARNING: Unable to add container CA certificates to JVM because %s is read-only", path)
 		return nil
 	}
 
-	out, err := os.OpenFile(c.KeyStorePath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
+	out, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
 	if err != nil {
-		return fmt.Errorf("unable to open %s\n%w", c.KeyStorePath, err)
+		return fmt.Errorf("unable to open %s\n%w", path, err)
 	}
 	defer out.Close()
 
-	if err := keystore.Encode(out, ks, []byte(c.KeyStorePassword)); err != nil {
+	if err := keystore.Encode(out, ks, []byte(password)); err != nil {
 		return fmt.Errorf("unable to encode keystore\n%w", err)
 	}
 
 	return nil
 }
+
+func CertificateLoaderMetadata(certificateLoader CertificateLoader) (map[string]interface{}, error) {
```

Comment:
Any reason this isn't a method on the `CertificateLoader`?

Reply:
A composition style, but I don't have a strong opinion either way.